packages/rest/src/router/metadata.ts

```diff
@@ -76,6 +77,57 @@ export function api(spec: ControllerSpec) {
   };
 }
 
+/**
+ * A decorator to mark a controller as belonging to a specific server.
+ * ```ts
+ * // In your application...
+ * class MyApplication extends Application {
+ *   constructor() {
+ *     super();
+ *     this.server(RestServer, 'foo');
+ *   }
+ * }
+ *
+ * // In one of your controllers
+ * @server('foo')
+ * class MyController {
+ *  // ...
+ * }
+ * ```
+ * @param name The name of the server bound to your application context.
+ */
+export function servers(names: string[]) {
+  return function(constructor: Constructor<any>) {
+    for (const name of names) {
+      // Set the value for this scope.
+      const endpoints = Reflector.getOwnMetadata(ENDPOINTS_KEY, constructor);
+      if (endpoints) {
+        rescopeMetadata(ENDPOINTS_KEY, name, constructor, endpoints);
```

Comment:
I would rather not re-scope here. The `@servers` can the following instead:

1. If applied to a method, define a metadata slot under `CONTROLLER_SCOPES_BY_METHOD_KEY', create or add an entry to the map: {*methodName*: [servers]} as the value
2. If applied to a class, define a metadata slot under `CONTROLLER_SCOPES_KEY` and use `servers` as the value

Then we can have helper function such as getControllerSepcByServer(serverName) to find matching specs.

Reply:
Are you saying you want the `@servers` decorator to be method-specific as well?
```ts
@servers(['RestServer', 'OtherRestServer'])
async getFoo(id: number) {
   // ???
}
```

Reply:
Yes, OpenAPI spec allows `servers` to be configured per operation. Here's what I expect to see after the decorators are applied:

```
MyController: 
CONTROLLER_SCOPES_KEY --> ['server1', 'server2']

MyController.prototype: 
CONTROLLER_SCOPES_BY_METHOD_KEY --> 
{
  method1: ['server1'],
  method2: ['server1', 'server2']
  // method3: undefined // Will be default to class level scopes
}
```

Then we should have a helper function to fetch the paths spec for a given server:
```
function getControllerSpecByServer(controllerClass, serverName: string) {
  // Build paths spec for the controller and its methods filtering by serverName
}
```

Please note that method level scopes depend on class level for defaults. But method level decorators are applied before class level ones. As a result, we cannot fully build the method level mapping inside `@servers` for a given method.

Reply:
Equating the way OpenAPI defines servers at their spec level with the "servers" concept we've created in LoopBack seems like a Bad Idea™.

Why is LoopBack as a microservice framework trying to fit directly into the OpenAPI v3 way of describing servers? Do we intend to read the "servers" collection and generate a number of RestServer instances based on that metadata? If so, why?

To me, this feels like a total misapplication of the OpenAPI spec, since I would **never** want to create multiple instances of things like environments (as per the OpenAPI example) in one deployable microservice. What do I do when my Dev environment is moving up to a new semver major and my Prod isn't yet ready for the same? Does this all still live in one module that we construct for the user based on the spec file?

The OpenAPI way of describing servers would be hosted at some higher level architecturally; an API management-level description of which catalogs contain the API as described.

In contrast, the "servers" decorator here allows code reuse so that developers don't have to write "base" controllers and extend them with wrapper classes just to decorate them for different targets.

IMO, multi server support is starting to feel more like an anti-pattern in general.
cc @bajtos

Reply:
Based on our discussion yesterday, my understanding is that we want multiple REST servers in the same process in order to support "public" and "operations" APIs - the first one being used by consumers of our API, the second used by our infrastructure for monitoring, health checks, etc.

To be honest, I am personally still not convinced that we need the complexity of generic support for arbitrary number of REST server instances in order to satisfy such requirement, but if the rest of the team believes this is the Right Way™, then I'll yield to unblock our progress.